ThreadPool no longer creates new threads to compensate for threads sleeping on futures

michaeljbishop opened the pull request:
This is an interesting, but still experimental version of the ThreadPool in
which the thread pool no longer needs to create new threads to
compensate for the imminent sleep of a thread on a future.

Previously, the ThreadPool created futures and added them to the queue.
If, when they processed a future and they had to wait for it to
finish executing, they would spawn another thread so they could
wait for the future.

Now, a non-blocking attempt to process the future's block is what is
put on the queue and a blocking attempt is passed back from #future.
In this way, the thread pool is free to process the core of a future
by dequeuing it and calling it. The the thread can't get the lock, it
is assumed that another thread is handling it and so it moves to
the next.

michaeljbishop committed:
Added some instrumentation to retieve stats from the threadpool

Two new methods to ThreadPool:

1 - #__begin_stats__
This turns on stat-keeping. You can only call it once.

2 - #__stats__
Retrieves all the current statistics. Locks the stats array while it
makes a duplicate. Best to call after #join.

michaeljbishop committed:
Merge branch 'master' of https://github.com/jimweirich/rake into no-sleep-thread-pool

Conflicts:
	lib/rake/thread_pool.rb

michaeljbishop committed:
Merge branch 'master' into no-sleep-thread-pool

michaeljbishop committed:
ThreadPool threads no longer sleep on futures.

Previously, the ThreadPool created futures and added them to the queue.
If, when they processed a future and they had to wait for it to
finish executing, they would spawn another thread so they could
wait for the future.

Now, a non-blocking attempt to process the future's block is what is
put on the queue and a blocking attempt is passed back from #future.
In this way, the thread pool is free to process the core of a future
by dequeuing it and calling it. The the thread can't get the lock, it
is assumed that another thread is handling it and so it moves to
the next.

jimweirich committed:
add private_reader support

jimweirich committed:
Merge branch 'master' of https://github.com/michaeljbishop/rake into threadstats

* 'master' of https://github.com/michaeljbishop/rake:
  Sped up #stat sampling.
  Fixed the ThreadPool statistics code to maintain compatibility with 1.8.7.

jimweirich committed:
move dot to end of previous line

to maintain compatibility with Ruby 1.8

jimweirich committed:
made the test less order dependent

jimweirich committed:
Merge branch 'no-sleep-thread-pool' of https://github.com/michaeljbishop/rake into newpool

* 'no-sleep-thread-pool' of https://github.com/michaeljbishop/rake:
  ThreadPool threads no longer sleep on futures.
  Added some instrumentation to retieve stats from the threadpool

jimweirich committed:
improve order independence in test

jimweirich committed:
add promise_complete? method

jimweirich committed:
whitespace

jimweirich committed:
more with promise_complete?

jimweirich committed:
should use value for futures (not call)

jimweirich committed:
add comment to assert

jimweirich committed:
add ThreadPool::Promise

jimweirich committed:
de-lambda-fied the promise object.

jimweirich commented:
I've locally merged the new pool changes.  In going over the code, I really felt that the promise should be its own object instead of a bunch of lambdas.  I created a Promise object, originally just to contain the lambdas, but as I kept refactoring the lambdas eventually melted away.

I'm fairly happy with the result and feel I have a better grasp on how the threading / future / promise model is working.  

The result is in the newpool branch.  Take a look.  If everything seems ok, I will merge this into master.

michaeljbishop commented:
That makes a lot of sense considering the promise was inspired by an actual Promise class.

One change that I'd like to make is to change what is added to the ThreadPool queue. If we instead add a lambda which executes the promise.work, then threads in the pool can bail early (since they didn't retrieve the lock, they can assume someone else is doing the work) and get started on the next thing.

The promise is always returned to the caller and will always sleep during #value.

I like the name #chore a lot!

On Nov 3, 2012, at 10:13 AM, Jim Weirich notifications@github.com wrote:

> I've locally merged the new pool changes. In going over the code, I really felt that the promise should be its own object instead of a bunch of lambdas. I created a Promise object, originally just to contain the lambdas, but as I kept refactoring the lambdas eventually melted away.
> 
> I'm fairly happy with the result and feel I have a better grasp on how the threading / future / promise model is working.
> 
> The result is in the newpool branch. Take a look. If everything seems ok, I will merge this into master.
> 
> —
> Reply to this email directly or view it on GitHub.

michaeljbishop commented:
heh. whoops. didn't read the code very carefully. That's basically what it does.

On Nov 3, 2012, at 10:13 AM, Jim Weirich notifications@github.com wrote:

> I've locally merged the new pool changes. In going over the code, I really felt that the promise should be its own object instead of a bunch of lambdas. I created a Promise object, originally just to contain the lambdas, but as I kept refactoring the lambdas eventually melted away.
> 
> I'm fairly happy with the result and feel I have a better grasp on how the threading / future / promise model is working.
> 
> The result is in the newpool branch. Take a look. If everything seems ok, I will merge this into master.
> 
> —
> Reply to this email directly or view it on GitHub.

michaeljbishop committed:
Fixed small bug in ThreadHistoryDisplay where it made a default :data
value that should be a Hash.

michaeljbishop committed:
Changed statistics to read more like a sentence.

Example:

  220823  F attempting_lock_on   item_id:20
  225342  F did_execute          item_id:20

Added more statistics to capture more detail.

michaeljbishop committed:
Removed inaccurate statistic about a thread awake.

jimweirich commented:
Re: chore ... Yes, I originally called it core for the lambda, but at some point the name slipped out.

I'm away from my laptop at the moment, I'll review the changes later tonight.

Quick question about thread life time ... Looks like the threads voluntarily die if the queue is empty. This works because adding a new item to the queue tries to start a new thread. First, is my understanding correct? If so. What do you think about keeping the thread around until the entire job is done, rather than trying to shrink the pool? Currently the code seems to have the potential to shrink/grow/shrink/grow.  Just thinking out loud, and maybe that's not a big concern.  Thoughts welcome.

michaeljbishop commented:
Re: thread death and resurrection

Yes, your understanding is correct. I don't have a strong argument for that behavior except that it makes it very convenient to join the pool and ensure all the work is complete. I don't otherwise have a good strategy for signaling the threads to quit.

However, here are a couple strategies I can think of off the top of my head:

1 - Have the threads sleep indefinitely on the queue
2 - Change #join to:
  2.1 - Set #max_active_threads to 0
  2.2 - Pump the queue with @threads.count no-op promises, so they are guaranteed to wake up from sleeping on the queue.

I'm a little uncomfortable with this because if the queue is not yet empty and join is called, items currently on the queue have the potential to add more items, yet we've set #max_active_threads to 0 and the threads will exit before they get to those newly added items.

Another strategy might be to keep the current strategy we have but implement our own queue allowing us to make a timeout for how long a thread will wait to dequeue an item. In that way there would at least be a delay before the thread exits, increasing the chance that it would be reused.

So, I'm not against the idea at all, I just don't have a good strategy for implementing it. I'm definitely open to ideas.

Somewhat unrelated but one other thing I'd like to experiment with:

I still have one concern that limiting the overall number of threads while having them sleep on futures might starve the thread pool. One option would be to use the #work+#chore method that we have now, but if any thread calls Promise#value and the thread sleeps, start an auxiliary thread (like the first version) to ensure there is always work being done. It's a hybrid of "replace sleeping threads" with the sleep-reduction technique of #work+#chore.

I added a statistic in my change so we could measure how much time threads spend sleeping inside Promise#value. I'm hoping it's low :)

I also haven't given up thinking about an alternative execution engine that uses a consistent number of threads, but doesn't require building a frozen tree before traversal. I doubt I'd be able to implement and test something before you'd like to release 10.0, but since the ThreadPool is private, that leaves me some wiggle room to try an alternative in a later version.

On Nov 3, 2012, at 16:19 PM, Jim Weirich notifications@github.com wrote:

> Re: chore ... Yes, I originally called it core for the lambda, but at some point the name slipped out.
> 
> I'm away from my laptop at the moment, I'll review the changes later tonight.
> 
> Quick question about thread life time ... Looks like the threads voluntarily die if the queue is empty. This works because adding a new item to the queue tries to start a new thread. First, is my understanding correct? If so. What do you think about keeping the thread around until the entire job is done, rather than trying to shrink the pool? Currently the code seems to have the potential to shrink/grow/shrink/grow. Just thinking out loud, and maybe that's not a big concern. Thoughts welcome.
> 
> —
> Reply to this email directly or view it on GitHub.

jimweirich commented:
Hmm, just a note.  I was testing with Rubinius and discovered that TestRakeTestThreadPool#test_pool_join_empties_queue will occasionally hang when running under Rubinius.  I haven't investigated any further than that.  Care to take a look?

michaeljbishop commented:
Yep, I'll get right on that…

On Nov 9, 2012, at 5:15 PM, Jim Weirich notifications@github.com wrote:

> Hmm, just a note. I was testing with Rubinius and discovered that TestRakeTestThreadPool#test_pool_join_empties_queue will occasionally hang when running under Rubinius. I haven't investigated any further than that. Care to take a look?
> 
> —
> Reply to this email directly or view it on GitHub.

jimweirich commented:
The deadlock problem may be an issue with Rubinius.  I've updated to the current head and haven't had a deadlock since then.  There was a deadlock issue that was addressed in https://github.com/rubinius/rubinius/commit/03a475f1d65f6ec2ee9c49dabcf322d22f1e81f4.

So, if you're not seeing anything, that might be why.

jimweirich commented:
I just ran 50 itereations of the tests with the latest master of Rubinius (with commit 52eba5f).  I got no crashes and no deadlocks.

michaeljbishop commented:
Hi Jim,

I tried using Rubinius and had the same result as you with no problems. I tried some other rubies and here are my results (on MOSX 10.8)

I ran the tests using:

  ruby -I lib bin/rake test

These rubies passed:

rubinius 2.0.0rc1 (1.8.7 52eba5f9 2012-11-02 JI) [x86_64-apple-darwin12.2.0]

jruby 1.6.7.2 (ruby-1.8.7-p357) (2012-05-01 26e08ba) (Java HotSpot(TM) 64-Bit Server VM 1.6.0_37) [darwin-x86_64-java]

ruby 1.9.3p194 (2012-04-20 revision 35410) [x86_64-darwin12.0.0]

ree-1.8.7-2012.02

These rubies crashed (though I am not sure why)

MacRuby 0.12 (ruby 1.9.2) [universal-darwin10.0, x86_64]
  Segmentation fault: 11

maglev 1.0.0 (ruby 1.8.7) (2011-10-31 rev 1.0.0-27184)[Darwin x86_64]

rake aborted!
NoMethodError: undefined method `length' for NilClass
/Users/mbishop/Dropbox/dev/github/jimweirich/rake/rakefile:195
(See full trace by running task with --trace)
a RubySystemExit occurred (error 2752)
topaz 1> exit

On Nov 10, 2012, at 17:27 PM, Jim Weirich notifications@github.com wrote:

> I just ran 50 itereations of the tests with the latest master of Rubinius (with commit 52eba5f). I got no crashes and no deadlocks.
> 
> —
> Reply to this email directly or view it on GitHub.
